lib/cosmos/dart/lib/dart_packet_log_writer.rb

```diff
@@ -0,0 +1,103 @@
+require File.expand_path('../../config/environment', __FILE__)
+require 'dart_common'
+
+class DartPacketLogWriter < Cosmos::PacketLogWriter
+  include DartCommon
+
+  DEFAULT_SYNC_COUNT_LIMIT = 100
+
+  def initialize(*args)
+    super(*args)
+    @packet_log_id = nil
+    @meta_id = nil
+    @db_queue = Queue.new
+    @sync_count = 0
+    @sync_count_limit = DEFAULT_SYNC_COUNT_LIMIT
+    @not_ready_ple_ids = []
+
+    sync_targets_and_packets()
+
+    @db_thread = Cosmos.safe_thread("Database packet log entries") do
+      db_thread_body()
+    end
+  end
+
+  def shutdown
+    super()
+    Cosmos.kill_thread(self, @db_thread)
+  end
+
+  def graceful_kill
+    super()
+    @db_queue << nil
+  end
+
+  protected
+
+  def start_new_file_hook(packet)
+    packet_log = PacketLog.new
+    packet_log.filename = @filename.clone
+    if @log_type == :TLM
+      packet_log.is_tlm = true
+    else
+      packet_log.is_tlm = false
+    end
+    packet_log.save!
+    @packet_log_id = packet_log.id
+    super(packet)
+  end
+
+  def pre_write_entry_hook(packet)
+    @sync_count += 1
+    if @sync_count > @sync_count_limit
+      @file.fsync
```

Comment:
Periodically syncing to the file system so that the data can be used by other Dart processes.

Reply:
After running I see a dart_cmd and dart_tlm file being created in outputs/dart/data. I assume this is what is flushing these. What is going into these files? Why do they exist?

Reply:
DART stores the raw packets for decommutation (and streaming) in regular COSMOS log files (which is what these are).  They are not stored directly the in SQL database.  SQL databases are really junk for binary data.  So instead they are stored in regular COSMOS files, and the database just keeps track of which packets are in the files and the offset in the file to each packet.  To get at the raw data, DART actually just reads it from a regular COSMOS log file.

Reply:
So right now we're double booking stuff: outputs/logs and dart/logs.

Reply:
It is double booking which isn't necessary bad.   I think the ideal setup will be to have DART running on a different computer and in that case you want double booking.  For the same computer, you can setup CmdTlmServer to use the dart_packet_log_writer and then nothing will be double booked.  Not sure of the performance of doing that though and for cases where you are not using JRuby that probably won't work well.     The files themselves don't actually differ in anyway except that the may start/stop at different points in time.

Reply:
Ok, we've been running our COSMOS bin archival tool on a different machine anyway so that's our typical use case. We push the files over and then DART just grabs everything in it's DART_DATA directory and processes it.

Reply:
For DART, you should have the CmdTlmServer streaming the data, and never have to copy any files except to handle data loss due to DART not running for some reason.